### PR TITLE
Add new API to set envar while specifying overwrite

### LINKED ISSUE #472: Option for rcutils_set_env() to not overwrite existing environment variables.
At present [rcutils_set_env](https://github.com/ros2/rcutils/blob/cf3e9986a557e4c893eb718b90c77ce189494a11/src/env.c#L50) always overwrites the value of the environment variable. However it would be nice to have the overwriting behavior configurable, ie, specify that `1` or `0` at the end of the `setenv` call.

This is useful in cases where the application would like to set the envar if previously unset in the environment.

### PR DESCRIPTION
**Pull request by Yadunund**
Fix #472

### CODE REVIEW

File: include/rcutils/env.h [hunk 1]
```diff
@@ -55,6 +55,24 @@ RCUTILS_WARN_UNUSED
 bool
```
Comment (Barry-Xu-2018) [inline include/rcutils/env.h:55]:
```suggestion
RCUTILS_DEPRECATED_WITH_MSG("Call rcutils_set_env_overwrite instead")
bool
```

File: include/rcutils/env.h [hunk 2]
```diff
@@ -55,6 +55,24 @@ RCUTILS_WARN_UNUSED
...
 rcutils_set_env(const char * env_name, const char * env_value);
 
+/// Set or un-set a process-scoped environment variable while specifying overwrite behavior.
+/**
+ * Documentation inherited from rcutils_set_env.
```
Comment (Barry-Xu-2018) [inline include/rcutils/env.h:60]: Move the detailed description of rcutils_set_env here, as the function will eventually be removed.

File: include/rcutils/env.h (hunk 1 above)
Comment (Yadunund) [reply include/rcutils/env.h:55]: Thanks for the feedback, should we call the new function `rcutils_set_env2`? (Similar to`rcl_timer_init2`)

File: src/env.c [hunk 3]
```diff
@@ -25,7 +25,7 @@ extern "C"
 #include "rcutils/error_handling.h"
 
 bool
-rcutils_set_env(const char * env_name, const char * env_value)
+rcutils_set_env_overwrite(const char * env_name, const char * env_value, bool overwrite)
```
Comment (fujitatomoya) [inline src/env.c:28]: `_putenv_s` does not have overwrite argument, so probably we need to call `getenv` to check if the environmental variable exists or not?

**Review by fujitatomoya — COMMENTED** [review]
I am okay to have this overwrite flag, but IMO user application most likely checks the environmental variable after this call to make sure if that is modified or not with overwrite flag is disabled. For me it would be more straight-forward to do, call getenv and then decide if setenv should be called or not in the application. maybe i am missing some use cases for that...

File: src/env.c (hunk 3 above)
Comment (mjcarroll) [reply src/env.c:28]: Yes, Windows annoyingly doesn't have this.  I ran into it when implementing the equivalent for Gazebo.  Generally, I think you can assume that as long as the arguments are valid, then the overwrite will succeed.

In this case, I think there are also some logic bugs.

For Windows, if the env is already set and you call `set_env_overwrite` with `overwrite=false`, you will end up overwriting (silently).

Comment (Yadunund) [reply src/env.c:28]: Thanks for highlighting these points. I guess the user should rely on `rcutils_get_evn` to check if an envar is set and only then call `rcutils_set_env`? If so we can close this PR.

**Comment by cottsay** [comment]
This was discussed in the ROS 2 ticket triage meeting today.

There are concerns about using the POSIX-specific overwrite parameter when we'd need to essentially implement the behavior from scratch on Windows. To avoid the diverging code paths while still implementing the desired API, it was suggested that the function could be written using the mechanisms we're already using elsewhere to reproduce similar behavior to the POSIX overwrite semantics but reliably behave the same on all supported platforms.

**Comment by Yadunund** [comment]
Thanks @cottsay for the valuable feedback!

> it was suggested that the function could be written using the mechanisms we're already using elsewhere to reproduce similar behavior to the POSIX overwrite semantics but reliably behave the same on all supported platforms.

Could someone share a link to such code for my reference?

**Comment by clalancette** [comment]
> Could someone share a link to such code for my reference?

Looking around the ROS 2 codebase, it actually turns out that I don't think we do this anywhere.  That is at least partially because `setenv` itself is very infrequently used.

Nevertheless, to make this operate under the same semantics as the Unix `setenv`, I imagine we'd do something like:

```
bool
rcutils_set_env_overwrite(const char * env_name, const char * env_value, bool overwrite)
{
  ...
  if (getenv(env_name) != NULL) {
    if (overwrite == 0) {
      return true;
    }
  }

  int set_ret;
#ifdef _WIN32
  if (NULL == env_value) {
    env_value = "";
  }
  set_ret = _putenv_s(env_name, env_value);
#else
  set_ret = setenv(env_name, env_value, (int)overwrite);
#endif

  if (set_ret != 0) {
    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("setting environment variable failed: %d", errno);
    return false;
  }

  return true;
```

**Comment by ahcorde** [comment]
@Yadunund any updates here ? what do you think about @clalancette suggestions ?

**Comment by Yadunund** [comment]
I've applied @clalancette's suggestion to the implementation and @Barry-Xu-2018's recommendation to rename this function to `rcutils_set_env2` and deprecate `rcutils_set_env`.

**Review by ahcorde — CHANGES_REQUESTED** [review]
There are some deprecation warnings https://build.ros2.org/job/Rpr__rcutils__ubuntu_noble_amd64/31/gcc/new/

We should get rid of them

File: src/env.c [hunk 4]
```diff
@@ -25,38 +25,43 @@ extern "C"
...
 
+  if (getenv(env_name) != NULL) {
+    if (overwrite == 0) {
+      return true;
+    }
+  }
+
+  int set_ret;
 #ifdef _WIN32
   if (NULL == env_value) {
```
Comment (ahcorde) [inline src/env.c:43]: There is a segfault here

```cpp
ASSERT_TRUE(rcutils_set_env2("NEW_ENV_VAR", nullptr, true));
```

when `env_value` inside this function is null, I think we can move this check outside of the `ifdef`

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by Yadunund** [comment]
> rcutils_set_env_overwrite

I've updated this PR such that it adds another API called `rcutils_set_env_overwrite` without deprecating `rcutils_set_env`. Also included a test.

**Review by ahcorde — APPROVED** [review]

**Comment by ahcorde** [comment]
Pulls: ros2/rcutils#473
Gist: https://gist.githubusercontent.com/ahcorde/fc88a94c71e0976f560ad9caed82a442/raw/c532c395841074fe3420ac842f60553b83ec9cbe/ros2.repos
BUILD args:  --packages-up-to  rcutils --packages-above-and-dependencies rcutils
TEST args:  --packages-select rcutils --packages-above rcutils
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14565
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21842/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16112/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1345/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22551/)

**Review by clalancette — APPROVED** [review]

**Comment by ahcorde** [comment]
https://github.com/Mergifyio backport jazzy iron

**Pull request merged**